src/include/sof_api/audio/sink_api.h

```diff
@@ -0,0 +1,178 @@
+ *       expensive - may involve some data moving in memory, cache writebacks, etc.
+ *
+ */
+
+/**
+ * Retrieves size of free space available in sink (in bytes)
+ * return number of free bytes in buffer available to immediate filling
+ */
+static inline size_t sink_get_free_size(struct sof_sink *sink)
+{
+	return sink->ops->get_free_size(sink);
+}
+
+static inline enum sof_ipc_frame sink_get_frm_fmt(struct sof_sink *sink)
+{
+	return sink->audio_stream_params->frame_fmt;
+}
+
+static inline uint32_t sink_get_channels(struct sof_sink *sink)
+{
+	return sink->audio_stream_params->channels;
+}
+
+/** get size of a single audio frame (in bytes) */
+static inline size_t sink_get_frame_bytes(struct sof_sink *sink)
+{
+	return get_frame_bytes(sink_get_frm_fmt(sink), sink_get_channels(sink));
+}
+
+/**
+ * Retrieves size of free space available in sink (in frames)
+ * return number of free frames in buffer available to immediate filling
+ */
+static inline size_t sink_get_free_frames(struct sof_sink *sink)
+{
+	return sink_get_free_size(sink) / sink_get_frame_bytes(sink);
+}
+
+/**
+ * Get a circular buffer to operate on (to write).
+ *
+ * Retrieves a fragment of circular data to be used by the caller
+ * After calling get_buffer, the space for data is guaranteed to be available
+ * for exclusive use on the caller core through provided pointer
+ * if the provided pointers are cached, it is guaranteed that the caller may safely use it without
+ * any additional cache operations
+ * The caller MUST take care of data circularity based on provided pointers
+ *
+ * @param sink a handler to sink
+ * @param [in] req_size requested size of space
+ * @param [out] data_ptr a pointer to the space will be provided there
+ * @param [out] buffer_start pointer to circular buffer start
+ * @param [out] buffer_size size of circular buffer
+ *
+ * @retval -ENODATA if req_size is bigger than free space
+ *
+ */
+static inline int sink_get_buffer(struct sof_sink *sink, size_t req_size,
+				  void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	int ret;
+
+	if (sink->requested_write_frag_size)
+		return -EBUSY;
+
+	ret = sink->ops->get_buffer(sink, req_size, data_ptr, buffer_start, buffer_size);
+
+	if (!ret)
+		sink->requested_write_frag_size = req_size;
+	return ret;
+}
+
+/**
+ * Commits that the buffer previously obtained by get_buffer is filled with data
+ * and ready to be used
+ *
+ * @param sink a handler to sink
+ * @param commit_size amount of data that the caller declares as valid
+ *	  if commit_size is bigger than the amount of data obtained before by get_buffer(), only
+ *	  the amount obtained before will be committed. That means - if somebody obtained a buffer,
+ *	  filled it with data and wants to commit it in whole, it may simple call
+ *	  commit_buffer with commit_size==MAXINT
+ * @return proper error code (0  on success)
+ */
+static inline int sink_commit_buffer(struct sof_sink *sink, size_t commit_size)
+{
+	int ret;
+
```

Comment:
Isn't it too much to inline?

Reply:
> Isn't it too much to inline?

`inline` is required in .h files to avoid the One Definition Rule:

https://stackoverflow.com/questions/39652884/one-definition-rule-multiple-definition-of-inline-functions
https://en.wikipedia.org/wiki/Inline_function

Every time I checked, the _optimizer_ did not care at all about `inline`.

Reply:
@marc-hb I think what @marcinszkudlinski meant is that if the function is too large to be inlined, it could be placed as a normal exported function in a .c file instead

Reply:
This function was moved to src/module/audio/sink_api.c